### PR TITLE
[MRG] Fixed early stopping mechanism

### LINKED ISSUE #55: Write tests for the held out validation data and early stopping
The validation data and early stopping mechanisms are not covered by the tests.

### PR DESCRIPTION
**Pull request by NicolasHug**
Closes #55 

- [X] add test coverage for early stopping mechanism
- [X] move verbose logic out of `_check_early_stopping`. Fixes iteration time computation
- [X] Early stopping is now checked at the end of the iteration, and not at the beginning. This fixes an issue where the first init predictions would be considered as a proper iteration.
- [X] Fixes the handling of `scoring` parameter. See comment in the code
- [X] Renames `max_no_improvement` to `n_iter_no_change`
- [X] Simplified early stopping logic to *early stop iff the current score doesn't improve any of the last `n_iter_no_change` by `tol` percent*

In scikit learn early stopping is done iff `n_iter_no_change` is not None. Here we do early stopping iff `scoring` is not None (there is no `scoring` in scikit learn).

### CODE REVIEW

File: pygbm/gradient_boosting.py [hunk 1]
```diff
@@ -141,154 +141,235 @@ def fit(self, X, y):
...
+            best = max(context_scores)
+            best_with_tol = best * (1 - tol * scorer._sign)
+            print()
+            print('candidate', candidate)
+            print('best_with_tol', best_with_tol)
+            print()
+            return candidate >= best_with_tol
+
+        # TODO: make sure that self.predict can work on binned data and
+        # then only use the public scorer.__call__.
```
[the remark this answers is not in the document]
Comment (ogrisel) [reply pygbm/gradient_boosting.py]: > @ogrisel is this even possible? How would we detect in self.predict(X) that X is pre-binned?

We could hardcode that if `X.dtype` is `np.uint8`, this is pre-binned data.

Alternatively, we introduce a new method (`predict_binned`).

For the other questions, I need to think about it more. I am not sure what is the solution either.

**Comment by NicolasHug** [comment]
Hm tests are passing but take too long without numba so travis times out.

Should we remove the `NUMBA_DISABLE_JIT="1"` jobs?

Anyway, ready for reviews :)

**Comment by ogrisel** [comment]
> Should we remove the NUMBA_DISABLE_JIT="1" jobs?

No because this is necessary to compute coverage. However we could skip the slowest test when `int(os.environ.get("NUMBA_DISABLE_JIT", 0)) == 1`.

**Comment by NicolasHug** [comment]
I see. Skipping this test when `NUMBA_DISABLE_JIT="1"` would result in much less coverage: this is the only test that covers early stopping and verbose=1 I think.

Should we have a dummy test that we don't disable, just for the coverage?

**Comment by ogrisel** [comment]
Yeah!

**Pull request merged**